pasta/base/token_generator.py

```diff
@@ -104,11 +104,38 @@ def predicate(token):
         self._loc = tok[2]
 
     # Eat a single newline character
-    if next_token[0] in (TOKENS.NL, TOKENS.NEWLINE):
+    if next_token and next_token[0] in (TOKENS.NL, TOKENS.NEWLINE):
       result += self.next()[1]
 
     return result
 
+  def block_whitespace(self, indent_level):
+    """Parses whitespace from the current _loc to the end of the block."""
+    # Get the normal suffix lines, but don't advance the token index unless
+    # there is no indentation to account for
+    starti = self._i
+    full_whitespace = self.whitespace()
+    if not indent_level:
+      return full_whitespace
+    self._i = starti
+
+    # Trim the full whitespace into only lines that match the indentation level
+    lines = full_whitespace.splitlines(True)
+    try:
+      last_line_idx = next(i for i, line in reversed(list(enumerate(lines)))
+                           if line.startswith(indent_level + '#'))
+    except StopIteration:
+      # No comment lines at the end of this block
+      self._loc = self._tokens[self._i][3]
+      return ''
+    lines = lines[:last_line_idx + 1]
+
+    # Advance the current location to the last token in the lines we've read
+    end_line = self._tokens[self._i][3][0] + 1 + len(lines)
+    list(self.takewhile(lambda tok: tok[2][0] < end_line))
+    self._loc = self._tokens[self._i][3]
```

Comment:
maybe you should have constants for these indices into the tokens to make this more readable?

Reply:
Good suggestion. Will do this in a followup.